api/v1alpha1/statuscheck_types.go

```diff
@@ -0,0 +1,215 @@
+// Copyright Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +chaos-mesh:base
+type StatusCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the behavior of a status check
+	Spec StatusCheckSpec `json:"spec"`
+
+	// +optional
+	// Most recently observed status of status check
+	Status StatusCheckStatus `json:"status"`
+}
+
+type StatusCheckMode string
+
+const (
+	// StatusCheckSynchronous means the status check will exit
+	// immediately after success or failure.
+	StatusCheckSynchronous StatusCheckMode = "Synchronous"
+	// StatusCheckContinuous means the status check will continue to
+	// execute until the duration is exceeded or the status check fails.
+	StatusCheckContinuous StatusCheckMode = "Continuous"
+)
+
+type StatusCheckType string
+
+const (
+	TypeHTTP StatusCheckType = "HTTP"
+)
+
+type StatusCheckSpec struct {
+	// Mode defines the execution mode of the status check.
+	// Support type: Synchronous / Continuous
+	// +optional
+	// +kubebuilder:validation:Enum=Synchronous;Continuous
+	Mode StatusCheckMode `json:"mode,omitempty"`
+
+	// Type defines the specific status check type.
+	// Support type: HTTP
+	// +kubebuilder:default=HTTP
+	// +kubebuilder:validation:Enum=HTTP
+	Type StatusCheckType `json:"type"`
+
+	// Duration defines the duration of the status check.
+	// A duration string is a possibly signed sequence of
+	// decimal numbers, each with optional fraction and a unit suffix,
+	// such as "300ms", "-1.5h" or "2h45m".
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// +optional
+	Duration *string `json:"duration,omitempty" webhook:"Duration"`
```

Comment:
Updated.

Reply:
It confuse me: if `Duration` could works with  `Synchronous`, it seems that the `Synchronous` status check would not be like an "one-shot check", am I correct?

Reply:
Maybe I need to modify the comment again, from `the duration of the whole status check if the execution is always successful` to `the duration of the whole status check if the failure threshold is not exceed`.

I think the `Duration` with `Continuous` is easy to understand, so let me explain the `Duration` with `Synchronous`. Here is an example:
```yaml
mode: Synchronous
duration: 12s
timeoutSeconds: 5
intervalSeconds: 5
failureThreshold: 3
```

1. If the first execution is successful, then this status check is successful and exits.
2. If the first execution fails because a timeout(5s), the status check should wait 5s and then trigger the next execution. And the `Failed` condition should be `unknown` because it's neither a `success` nor a `failure` (not exceed the failure threshold).
   - if the second execution is successful in 2s, then this status check is successful and exits.
   - if the second execution is not successful in 2s, the first execution(5s) + interval(5s) + the second execution(2s) are already exceed the duration `12s`, so the status check should exit, and the `Failed` condition keeps `unknown` status.

Reply:
> the Synchronous status check would not be like an "one-shot check"

The Synchronous status check will exit once it succeed, but it will retry when it fails, and it will exit until it exceeds the failure threshold (as a failure status check). So the Synchronous could have multi-executions, but it will exit when it confirms a success or a failure.

Reply:
Oh, I get your point. So with StatusCheck with a small duration (in detail `duration < failureThreshold * intervalSeconds`), the `Failed` would never be `True`? 🤔

- Is this expected?
- Does it make sense?

Reply:
I agree with you. :) But it is still a little counter-intuitive.

I think we should describe this in our document, and we could give a warning message if there is a possibility for making the condition `Failed` never be `True`(in other words, not possible to stop the related workflow).

What do you think about it? @iguoyr  @YangKeao @WangXiangUSTC @cwen0

Reply:
Maybe we could check it in Validate webhook? 🤔 

`duration` > `failureThreshold` * (`timeoutSeconds` + `intervalSeconds`)

Reply:
> Maybe we could check it in Validate webhook? thinking
> 
> `duration` > `failureThreshold` * (`timeoutSeconds` + `intervalSeconds`)

I am not sure about that. 

What is the purpose of the field `duration`? If I understand correctly, be similar to Chaos, `duration` in StatusCheck also means "this durable action should NOT run longer than `duration`". 

I think the length of `duration` might be determined by the usage of different situations and has no relation with the `failureThreshold`, `timeoutSeconds`, and `intervalSeconds`. So a `small` duration just makes sense. 

I prefer to only warn or notice that somewhere instead of making the validation rule.

Reply:
@STRRL Updated! PTAL

Reply:
The latest update is awesome!